drivers/rf-transceiver/hmc6300/hmc6300.c

```diff
@@ -0,0 +1,439 @@
+error:
+	free(d->vco.freqs);
+	return ret;
+}
+
+int hmc6300_init(struct hmc6300_dev **dev, struct hmc6300_init_param *init)
+{
+	struct hmc6300_dev *d;
+	int ret;
+
+	if (!dev || !init)
+		return -EINVAL;
+
+	d = (struct hmc6300_dev *) calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	switch(init->ref_clk) {
+	case HMC6300_REF_CLK_71p42857MHz:
+		ret = _hmc6300_generate_vco(d, 56500000000, 250000000, 0x22, 31);
+		break;
+	case HMC6300_REF_CLK_75MHz:
+		ret = _hmc6300_generate_vco(d, 55125000000, 262500000, 0x12, 43);
+		break;
+	case HMC6300_REF_CLK_142p8571MHz:
+		ret = _hmc6300_generate_vco(d, 56500000000, 500000000, 0x11, 16);
+		break;
+	case HMC6300_REF_CLK_154p2857MHz:
+		ret = _hmc6300_generate_vco(d, 57240000000, 540000000, 0x0a, 17);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->en, &init->en);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->clk, &init->clk);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->data, &init->data);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->scanout, &init->scanout);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->en, NO_OS_GPIO_HIGH);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->clk, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->data, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_input(d->scanout);
+	if (ret)
+		goto error;
+
+	ret = hmc6300_write_regmap(d, hmc6300_default_regmap);
+	if (ret)
+		goto error;
+
+	*dev = d;
+	return 0;
+error:
+	free(d);
+	return ret;
+}
+
+int hmc6300_remove(struct hmc6300_dev *dev)
+{
+	free(dev->vco.freqs);
+	free(dev->vco.fbdiv);
+	no_os_gpio_remove(dev->en);
+	no_os_gpio_remove(dev->clk);
+	no_os_gpio_remove(dev->data);
+	no_os_gpio_remove(dev->scanout);
+	free(dev);
+	return 0;
+}
+
+/* Write a device row using GPIO bit-banging. */
+int hmc6300_write_row(struct hmc6300_dev *dev, uint8_t row, uint8_t val)
+{
+	int b;
+	uint32_t send = 0;
+	send = val; // data
+	send |= ((uint32_t)row & 0x3f) << 8; // array address
+	send |= 1 << 14; // write
+	send |= ((uint32_t)0x6) << 15;  // chip address
+	const int delay = 1;
+
+	no_os_gpio_set_value(dev->en, NO_OS_GPIO_LOW);
+	for (b = 0; b < 18; b++) {
+		no_os_gpio_set_value(dev->data, send & 0x1);
+		no_os_udelay(delay);
+		no_os_gpio_set_value(dev->clk, NO_OS_GPIO_HIGH);
+		no_os_udelay(delay);
+		no_os_gpio_set_value(dev->clk, NO_OS_GPIO_LOW);
+		send >>= 1;
+	}
+	no_os_gpio_set_value(dev->data, NO_OS_GPIO_LOW);
+	no_os_gpio_set_value(dev->en, NO_OS_GPIO_HIGH);
+
+	return 0;
+}
+
+/* Read a device row using GPIO bit-banging. */
+int hmc6300_read_row(struct hmc6300_dev *dev, uint8_t row, uint8_t *val)
+{
+	int b;
+	uint8_t recv = 0;
+	uint32_t send = 0;
+	send |= ((uint32_t)row & 0x3f) << 8;
+	send |= ((uint32_t)0x6) << 15;
+	const int delay = 1;
+
+	// write the first 18 bits on data
+	no_os_gpio_set_value(dev->en, NO_OS_GPIO_LOW);
```

Comment:
I think here it's acceptable to not check the return because it can only be 0 or -EINVAL, and we know through code review that we're passing valid parameters to set_value function. I will add an in-code comment about this.